python/eight_mile/pytorch/layers.py

```diff
@@ -1718,6 +1720,91 @@ def _attention(self, query: torch.Tensor, key: torch.Tensor, mask: Optional[torc
+    """
+    def __init__(self, hsz: int = None, pdrop: float = 0.1, **kwargs):
+        super().__init__()
+        self.hsz = hsz
+        self.dropout = nn.Dropout(pdrop)
+        self.attn = None
+
+    def forward(self, qkvrm: Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]) -> torch.Tensor:
+        query, key, value, edges_key, edges_value, mask = qkvrm
+        a = self._attention(query, key, edges_key, mask)
+        self.attn = a
+        a = self.dropout(a)
+        return self._update(a, value, edges_value)
+
+    def _attention(self, query: torch.Tensor, key: torch.Tensor, edges_key: torch.Tensor, mask: Optional[torch.Tensor] = None) -> torch.Tensor:
+        pass
+
+    def _update(self, a: torch.Tensor, value: torch.Tensor, edges_value: torch.Tensor) -> torch.Tensor:
+        """Attention weights are applied for each value, but in a series of efficient matrix operations.
+
+        In the case of self-attention, the key and query (used to create the attention weights)
+        and values are all low order projections of the same input.
+
+        :param a: The attention weights [B, H, T, T]
+        :param value: The values [B, H, T, D]
+        :param edge_value: The edge values [T, T, D]
+        :returns: A tensor of shape [B, H, T, D]
+        """
+        B, H, T, D = value.shape
+        updated_values = torch.matmul(a, value)
+        a = a.view(B*H, T, T).transpose(0, 1)  # (T, BxH, T)
+        t = torch.matmul(a, edges_value)  # (T, BxH, D)
+        update_edge_values = t.transpose(0, 1).view(B, H, T, D)
+        return updated_values + update_edge_values
+
+
+class SeqScaledDotProductRelativeAttention(SequenceSequenceRelativeAttention):
+    def __init__(self, pdrop: float = 0.1, **kwargs):
+        super().__init__(pdrop=pdrop, **kwargs)
+
+    def _attention(self, query: torch.Tensor, key: torch.Tensor, edges_key: torch.Tensor, mask: Optional[torch.Tensor] = None) -> torch.Tensor:
+        """Scaled dot product attention, as defined in https://arxiv.org/abs/1706.03762
+
+        We apply the query to the keys to receive our weights via softmax in a series of efficient
+        matrix operations. In the case of self-attntion the key and query are all low order
+        projections of the same input.
+
+        :param query: a query for alignment. Can come from self in case of self-attn or decoder in case of E/D
+        :param key: a set of keys from encoder or self
+        :param mask: masking (for destination) to prevent seeing what we shouldnt
+        :param edges_key: a matrix of relative embeddings between each word in a sequence [TxTxD]
+        :return: A tensor that is (BxHxTxT)
+        """
+        # (., H, T, T) = (., H, T, D) x (., H, D, T)
+        B, H, T, d_k = query.shape
+        scores_qk = torch.matmul(query, key.transpose(-2, -1))
+        tbhd = query.reshape(B*H, T, d_k).transpose(0, 1)
+        scores_qek = torch.matmul(tbhd, edges_key.transpose(-2, -1))
```

Comment:
edges key is `[T, T, d_k]` transposed to `[T, d_k, T]`
tbhd is `[T, B*H, d_k]`
the mult returns `[T, B*H, T]`
The transpose creates `[B*H, T, T]`
The reshapes gives us `[B, H, T, T]`

This seems to work and match the paper :+1: